numba/unicode.py

```diff
@@ -1286,6 +1288,137 @@ def impl(a):
+            x += 1
+        final_sigma = (x == l or (not _PyUnicode_IsCased(_Py_UCS4(c))))
+    ret = 0x3c2 if final_sigma == 1 else 0x3c3
+    return ret
+
+
+@register_jitable
+def _lower_ucs4(code_point, data, length, idx, mapped):
+    """
+    https://github.com/python/cpython/blob/201c8f79450628241574fba940e08107178dc3a5/Objects/unicodeobject.c#L9885-L9895
+    """
+    if code_point == 0x3A3:
+        mapped[0] = _handle_capital_sigma(data, length, idx)
+        return 1
+    return _PyUnicode_ToLowerFull(_Py_UCS4(code_point), mapped)
+
+
+@register_jitable
+def _do_upper_or_lower(data, length, res, maxchars, lower):
+    """
+    https://github.com/python/cpython/blob/201c8f79450628241574fba940e08107178dc3a5/Objects/unicodeobject.c#L9946-L9965
+    """
+    k = 0
+    for idx in range(length):
+        mapped = np.zeros(3, dtype=_Py_UCS4)
+        code_point = _get_code_point(data, idx)
+        if lower:
+            n_res = _lower_ucs4(code_point, data, length, idx, mapped)
+        else:
+            n_res = _PyUnicode_ToUpperFull(_Py_UCS4(code_point), mapped)
+        for m in mapped[:n_res]:
+            maxchar, = maxchars
+            maxchars[0] = max(maxchar, m)
+            _set_code_point(res, k, m)
+            k += 1
+    return k
+
+
+@register_jitable
+def _do_lower(data, length, res, maxchars):
+    """
+    https://github.com/python/cpython/blob/201c8f79450628241574fba940e08107178dc3a5/Objects/unicodeobject.c#L9973-L9977
+    """
+    return _do_upper_or_lower(data, length, res, maxchars, lower=True)
+
+
+@overload_method(types.UnicodeType, 'lower')
+def unicode_lower(data):
+    """
+    Implements .lower()
+    """
+    def impl(data):
+        """
+        main structure is a translation of:
+        https://github.com/python/cpython/blob/201c8f79450628241574fba940e08107178dc3a5/Objects/unicodeobject.c#L12380-L12388
+        """
+
+        # ASCII fast path
+        length = len(data)
+        if data._is_ascii:
+            # This is an approximate translation of:
+            # https://github.com/python/cpython/blob/201c8f79450628241574fba940e08107178dc3a5/Objects/bytes_methods.c#L247-L255
+            res = _empty_string(data._kind, length, data._is_ascii)
+            for idx in range(length):
+                code_point = _get_code_point(data, idx)
+                _set_code_point(res, idx, _Py_TOLOWER(code_point))
+            return res
+        else:
+            # This is an approximate translation of:
+            # https://github.com/python/cpython/blob/201c8f79450628241574fba940e08107178dc3a5/Objects/unicodeobject.c#L10023-L10069
+            tmp = _empty_string(PY_UNICODE_4BYTE_KIND, 3 * length, data._is_ascii)
+            # maxchar should be inside of a list to be pass as argument by reference
+            maxchar = 0
+            maxchars = [maxchar]
+            newlength = _do_lower(data, length, tmp, maxchars)
+            maxchar, = maxchars
+            newkind = _codepoint_to_kind(maxchar)
```

Comment:
I would suggest the following to shorten this.

```suggestion
            maxchars = [0]
            newlength = _do_lower(data, length, tmp, maxchars)
            newkind = _codepoint_to_kind(maxchar[0])      
```

Reply:
I will make the minor change.